metadata/src/from_into/v14.rs

```diff
@@ -307,70 +309,81 @@ impl ExtrinsicPartTypeIds {
 
 fn generate_outer_enums(
     metadata: &mut v14::RuntimeMetadataV14,
-) -> v15::OuterEnums<scale_info::form::PortableForm> {
-    let call_enum = metadata
-        .types
-        .types
-        .iter()
-        .find(|ty| {
+) -> Result<v15::OuterEnums<scale_info::form::PortableForm>, TryFromError> {
+    let mut path = None;
+
+    let mut find_type = |name: &str| {
+        metadata.types.types.iter().find_map(|ty| {
             let Some(ident) = ty.ty.path.ident() else {
-                return false;
+                return None;
             };
-            ident == "RuntimeCall"
-        })
-        .expect("RuntimeCall exists in V14; qed");
 
-    let event_enum = metadata
-        .types
-        .types
-        .iter()
-        .find(|ty| {
-            let Some(ident) = ty.ty.path.ident() else {
-                return false;
+            if ident != name {
+                return None;
+            }
+
+            let TypeDef::Variant(_) = &ty.ty.type_def else {
+                return None;
             };
-            ident == "RuntimeEvent"
+
+            if path.is_none() {
+                let mut segments = ty.ty.path.segments.clone();
+                segments.pop();
+                path = Some(segments);
+            }
```

Comment:
It's a bit weird saving a path mutably like this I think. More code but IMO clearer would be something like this:

```rust
// here, return eg:
Some((ty.id, segments))

// and then:
let call_enumm = find_type("RuntimeCall");
let event_enumm = find_type("RuntimeEvent");
let error_enumm = find_type("RuntimeError");

// Pick a path to save any generated types to; aim to
// put them next to whatever existing type we find first,
// or fall back to dumping it in the "top level" runtime_types
// path.
let path = match (call_enum, event_enum, error_enum) {
    (Some(p), None, None) => p,
    (None, Some(p), None) => p,
    (None, None, Some(p)) => p,
    (None, None, None) => vec!["runtime_types".into()]
};

// ...
```

Reply:
I'll remove that bit; fun fact I had initially thought of something similar with `call.or(event).or(error)` :D

Reply:
Ooh yeah you're right; using `.or` would have been much more compact :D